### DISCUSSION TITLE
How to obtain forecasts for different models for each horizon (direct reduction strategy)?

### DISCUSSION THREAD

**Discussion by hliebert** [alan-turing-institute/sktime#4884] [Q&A] [answered]
As part of a larger migration at work, I'm currently porting a forecasting model from R to Python using `sktime`. Results from the model are going into several reports, so in a first step I'm trying to replicate it as closely as possible (regardless of whether all of the design choices are sensible or not).

I've previously used sktime for small time series forecasts, but never for more complex composite models, so I'm trying to gauge how to put the different components together.

The model uses a direct reduction forecasting strategy, cross-validates different tabular regression models, returns a fixed number of n_best models/predictions for each model for each horizon, then selects the median among them.

I've looked at `ForecastingGridSearchCV` and `n_best_forecasters_` appears to return the best models over the whole horizon and only accepts a single forecaster. Can something like the above be composed, or should I just do grid search for each model and horizon independently, and then use `n_best_forecasters_` from each of these to obtain the predictions for each combination?

Edit: Possibly related to https://github.com/sktime/sktime/discussions/4584.

**Comment by fkiraly** [comment]
0.20.1 will have a forecaster `FhPlexForecaster`, which allow to specify different forecasters per forecasting horizon:

https://github.com/sktime/sktime/pull/4811

Using different models for different horizons was a frequent user request, so this might be what you are looking for?

FYI, it is the first draft so might be a bit rough around the edges - so testing and feedback (and contribution 😄 ) would be greatly appreciated.

PS: if you're doing the porting exercise now, it would be much appreciated and probably useful to you if you could either:

* fill out a user feedback/registration form before the community sprint on July 22-23 so your requirements go into the roadmap planning process. the form is here at the top (link): http://www.sktime.net/en/latest/
* join the community sprint on July 22-23 in-person in Prague, or virtually (see our linkedin for details)

**Reply by hliebert** [reply]
Thank you, `FhPlexForecaster` seems to be half of what I want. The other half would be something like `MultiplexForecaster`, where instead of cross-validating `selected_forecaster`, each forecaster is optimized independently. For now, I'll just use separate `FhPlexForecaster` instances for each model.  

I've filled out the feedback/registration form (and encouraged my colleagues to do the same).

**Reply by hliebert** [reply]
I've tried using `FhPlexForecaster`, but I'm not sure whether I'm using it correctly in combination with cross validation. I was expecting the following to work, but it returns the same model and parameters for each horizon. 

The only ordering that seems to work is if I nest `ForecastingRandomizedSearchCV` within `FhPlexForecaster`. From the documentation, I'm not sure what the intended usage is here. The example for `FhPlexForecaster` is not related and probably out of place.

```python
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sktime.datasets import load_longley
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.compose import make_reduction, FhPlexForecaster
from sktime.forecasting.model_selection import ExpandingWindowSplitter
from sktime.forecasting.model_selection import ForecastingRandomizedSearchCV
from sktime.performance_metrics.forecasting import MeanSquaredError

y, X = load_longley()
horizon = ForecastingHorizon(np.arange(1, 4), is_relative=True)

random_forest = RandomForestRegressor()
reduction_forecaster = make_reduction(random_forest, window_length=2, strategy="direct", windows_identical=False)

horizon_specific_forecaster = FhPlexForecaster(forecaster=reduction_forecaster)
horizon_specific_forecaster.get_params()

expanding_window_cv = ExpandingWindowSplitter(fh=horizon, step_length=1)
grid = {"forecaster__estimator__max_features": np.linspace(0.1, 0.9, num=20)}

cv_model = ForecastingRandomizedSearchCV(
    forecaster=horizon_specific_forecaster,
    cv=expanding_window_cv,
    param_distributions=grid,
    n_iter=3,
    scoring=MeanSquaredError(),
    strategy="refit",
    refit=True,
    return_n_best_forecasters=1,
    update_behaviour="full_refit",
    error_score="raise",
)

cv_model.fit(y, X, fh=horizon)
cv_model.get_fitted_params()
```

**Reply by fkiraly** [reply]
oops, it looks like a copy-paste monster ate the examples in the estimator documentation of `FhPlexForecaster`... I'll fix that and also post example code here.

**Reply by fkiraly** [reply]
Just looked at the code above - what this will do is fit a different `reduction_forecaster` per horizon, and the randomized grid search will optimize the parameters overall, i.e., use the same parameter for all the `reduction_forecaster`-s. Why? Because there's only one `reduction_forecaster` that the search sees "from the outside".

If you want different grid searches per horizon, you need to nest the other way round, i.e., put the grid search inside the `FhPlexForecaster`, i.e., `FhPlexForecaster(ForecastingRandomizedSearchCV(DirectReduction(RandomForest(...` rather than `ForecastingRandomizedSearchCV(FhPlexForecaster(DirectReduction(RandomForest(...`

I'll try to clarify the combination with grid search in the new examples.

**Reply by hliebert** [reply]
Thank you very much for taking the time to explain! That's what I thought - I am currently using it like this. For some reason I initially expected the cross validation to be the last step in the pipeline. 

I was unsure whether the usage was correct because the forecasts were looking off for horizons > 1, but I think I may have stumbled upon a bug there (the forecasts are only off if the forecaster passed to grid search is a pipeline which includes a `Differencer`. I'll open a separate bug report for that.

**Reply by fkiraly** [reply]
hm, I suspect it might have to do with the `fh` not bein a regular continuation of the data index?

If I'm guessing that's the case, perhaps put the differencer outside the `FhPlexForecster` for a quick workaround (but we still should fix the bug!)

**Reply by hliebert** [reply]
Thanks, I'll do that and file a report. For completeness I'm posting a rearranged variant of the example above that works.

```python
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sktime.datasets import load_longley
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.compose import make_reduction, FhPlexForecaster
from sktime.forecasting.model_selection import ExpandingWindowSplitter
from sktime.forecasting.model_selection import ForecastingRandomizedSearchCV
from sktime.performance_metrics.forecasting import MeanSquaredError

y, X = load_longley()
horizon = ForecastingHorizon(np.arange(1, 4), is_relative=True)

random_forest = RandomForestRegressor()
reduction_forecaster = make_reduction(random_forest, window_length=2, strategy="direct", windows_identical=False)

expanding_window_cv = ExpandingWindowSplitter(fh=horizon, step_length=1)
grid = {"estimator__max_features": np.linspace(0.1, 0.9, num=20)}

cv_model = ForecastingRandomizedSearchCV(
    forecaster=reduction_forecaster,
    cv=expanding_window_cv,
    param_distributions=grid,
    n_iter=3,
    scoring=MeanSquaredError(),
    strategy="refit",
    refit=True,
    return_n_best_forecasters=1,
    update_behaviour="full_refit",
    error_score="raise",
    verbose=1,
)

horizon_specific_forecaster = FhPlexForecaster(forecaster=cv_model)

horizon_specific_forecaster.fit(y, X, fh=horizon)
horizon_specific_forecaster.predict(fh=horizon, X=X)
```

**Comment by fkiraly** [comment]
> The other half would be something like `MultiplexForecaster`, where instead of cross-validating `selected_forecaster`, each forecaster is optimized independently.

Can you kindly explain this in more detail, @hliebert?

I think you can already build this, but I'm not sure what exactly you want to build.

If you want different forecasters per horizon, it works like this:

```python
multiplex = MyForecasterA(...) ¦ MyForecasterB(...)

# let's say we want to forecast 1 step ahead with A, and 2 steps ahead with B
fh_params = [{"selected_forecaster": "MyForecasterA"}, {"selected_forecaster": "MyForecasterB"}]

fhplex = FhPlexForecaster(multiplex, fh_params)
```

If you want to optimize the forecasters independently, replace `MyForecasterA` and B by a grid search pipeline or similar.

Noting, the optimization in this case can live both *inside* or *outside* the `FhPlexForecaster`.
If inside, the optimization will be for the metric computed only on the `fh` element. If outside, the optimization will be for the metric computed across all `fh` elements. There will be more grid/search elements in general for the same optimization space, if you do it outside with the naive grid/search protocol, rather than inside.

**Comment by hliebert** [comment]
Sure. The model I am trying to replicate is essentially doing something in the spirit of the above several times for different estimators, then combines the results. Basically

- Do cross-validation for k candidate models, independently for each horizon.
- For each estimator for each horizon, return the predictions of a fixed number of best models (say the 100 `n_best_forecasters_`). 
- Combine the predictions by horizon and and select e.g. the median for each horizon.

From my reading of `MultiPlexForecaster` I don't think it'll be of help here (but I may be wrong). Now I simply estimate one `FhPlexForecaster` for each candidate model. 

One reason for why I was originally trying to nest `FhPlexForecaster` within the cross validation is the duplication of the transformation pipeline (which necessarily needs to run within the cross validation). For example, if the transformation pipeline contains a feature selection model that takes a while to run, but is independent of the ultimate estimator (hence the pipeline is the same for every model), that effort is duplicated for every model if the cross-validation is nested within.

**Reply by fkiraly** [reply]
ok, that's an interesting combination!

You can achieve what you want to do afaik already with existing components as follows. Let's build only the two last bullet points, as we have already figured out how to do the first.

I.e., the recipe below will do:

* do cross-validation, get params for the `N_BEST` best models
* take the predictions from those `N_BEST` best models
* combine the predictions, e.g., by mean or median aggregation

The pattern to use is the "parameter plugin", you need to plug in the `n_best_forecasters_` fitted parameter into the ensembler, the `EnsembleForecaster`. You do this as follows:

```python
from sktime.forecasting.compose import EnsembleForecaster
from sktime.forecasting.model_selection import ForecastingGridSearchCV
from sktime.param_est.plugin import PluginParamsForecaster

# let's say this is the number you want to aggregate
N_BEST = 100

# let's say this is your grid searcher
gscv = ForecastingGridSearchCV(forecaster, cv, param_grid, return_n_best_forecasters=N_BEST)

# now we'll plug in the N best forecasters into the ensembler!
ensembler = EnsembleForecaster([], aggfunc="median")  #dummy at the parameter we will replace by plugin
plugin_fcst = PluginParamsForecaster(gscv, ensembler, {"n_best_forecasters": "forecasters"})
```

The `plugin_fcst` can be used as any other forecaster.

What `plugin_fcst` does in `fit`:

1. grid search of `forecaster` using `cv`, `param_grid` etc; stores the `N_BEST` forecasters in its `"n_best_forecasters"` fitted param
2. that fitted param's value is plugged into `EnsembleForecaster` `ensembler`'s `forecasters` parameter
3. a clone of that modified `ensembler` is then fitted on all data seen in `fit`.

In `predict` etc, the `plugin_fcst` behaves like the `ensembler` obtained from step 3 above.

**Reply by fkiraly** [reply]
> The `plugin_fcst` can be used as any other forecaster.

... in particular, you can stick it into `FhPlexForecaster` to get the first bullet point as well.

**Reply by hliebert** [reply]
Thank you! I wasn't aware of PluginParamsForecaster. This is much neater than what I had in mind. 

I've tried to apply it to the example above, but it doesn't seem to work. A check fails with `ValueError: Invalid 'estimators' attribute, 'estimators' should be a list of (string, estimator) tuples` (though that is exactly what `n_best_forecasters` contains).

```python
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sktime.datasets import load_longley
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.compose import make_reduction, EnsembleForecaster
from sktime.forecasting.model_selection import ExpandingWindowSplitter
from sktime.forecasting.model_selection import ForecastingGridSearchCV
from sktime.param_est.plugin import PluginParamsForecaster

y, X = load_longley()
horizon = ForecastingHorizon(np.arange(1, 4), is_relative=True)

random_forest = make_reduction(RandomForestRegressor(), window_length=2, strategy="direct", windows_identical=False)
expanding_window_cv = ExpandingWindowSplitter(fh=horizon, step_length=1)
cv_random_forest = ForecastingGridSearchCV(
    forecaster=random_forest,
    cv=expanding_window_cv,
    param_grid={"estimator__max_features": np.linspace(0.1, 0.9, num=5)},
    return_n_best_forecasters=3,
    verbose=1,
    n_jobs=6,
)

ensembler = EnsembleForecaster(forecasters=[], aggfunc="median")
plugin_fcst = PluginParamsForecaster(param_est=cv_random_forest, forecaster=ensembler, params={"n_best_forecasters": "forecasters"})

plugin_fcst.fit(y, X, fh=horizon)
plugin_fcst.predict(fh=horizon, X=X)

cv_random_forest.fit(y, X, fh=horizon)
cv_random_forest.get_fitted_params()["n_best_forecasters"]
```

**Reply by fkiraly** [reply]
Apologies, looks like this is a genuine bug.
(the parameter estimation module is maturing but not stable)

![image](https://github.com/sktime/sktime/assets/7985502/82ad0700-b252-4167-b7a3-3dd744a6c26d)

The bug caused that nothing was being plugged into the `ensembler`, due to an "error of sign" in parameter checking (the checking/subsetting was the other way round than the plugin indicated by the `dict`, which could lead to an empty dict in plugin).

Reported here:
https://github.com/sktime/sktime/issues/4921

Fixed here:
https://github.com/sktime/sktime/pull/4922

Testing and feedback appreciated.

Fix should be in the next version.

**Reply by hliebert** [reply]
That's why the ensembler was showing empty. Thanks for fixing it! I'll swap the dict arguments and test when available. 

Can I ask one more question? In the example above, you're plugging only one grid search estimator. Is it possible to tweak the ensemble such that it collects the `n_best_forecasters` from multiple grid search models?

Ideally I want to do the ensembling for different models (e.g. not just random forest). So basically collect `N_BEST` models from multiple cross-validated models each, and then extract the median from all of those predictions. Is there a way to extend `PluginParamsForecaster` to do that?

**Reply by fkiraly** [reply]
> Is it possible to tweak the ensemble such that it collects the `n_best_forecasters` from multiple grid search models?

Would that not be the same as giving the grid search a union grid? That is, `param_grid = [grid1, grid2]`?

> So basically collect N_BEST models from multiple cross-validated models each, and then extract the median from all of those predictions

That would be just an `EnsembleForecaster` of forecasters like above, no? So,

`EnsembleForecaster([("model1", PluginParams(GridSearch(...), EnsembleForecaster(...))), ("model2", PluginParams(GridSearch(...), EnsembleForecaster(...))`

etc - is this what you mean?

Side note: the amount of brackets in play indicates, this is a quite complex, computation-heavy, and nested model, you really want to test it carefully against simpler alternatives. Of course if the purpose is replicating state-of-art in a comparison study, that's also reason enough to build it.

**Reply by hliebert** [reply]
Sorry, I think I didn't express myself very well. The second is sort of what I'm trying to replicate, without nesting the multiple ensembles. Basically I want to stack predictions from different models and then select the median. I've added an example for illustration below.

You're right, this is complex (probably unnecessarily so) and computation-heavy. That is partly the reason for the migration. I just need to replicate it once as-is (or at least close enough), then I can evaluate alternatives and simplify it.

In the example, I'd want to stack the `N_BEST` forecasts from both `cv_linear_model` and `cv_random_forest` by horizon and then select the median for each horizon.

```python
import numpy as np
from sklearn.linear_model import ElasticNet
from sklearn.ensemble import RandomForestRegressor
from sktime.datasets import load_longley
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.compose import make_reduction, FhPlexForecaster, EnsembleForecaster
from sktime.forecasting.model_selection import ExpandingWindowSplitter
from sktime.forecasting.model_selection import ForecastingGridSearchCV
from sktime.param_est.plugin import PluginParamsForecaster

y, X = load_longley()
horizon = ForecastingHorizon(np.arange(1, 4), is_relative=True)

linear_model = make_reduction(ElasticNet(), window_length=2, strategy="direct", windows_identical=False)
random_forest = make_reduction(RandomForestRegressor(), window_length=2, strategy="direct", windows_identical=False)

expanding_window_cv = ExpandingWindowSplitter(fh=horizon, step_length=1)

cv_linear_model = ForecastingGridSearchCV(
    forecaster=linear_model,
    cv=expanding_window_cv,
    param_grid={"estimator__alpha": np.linspace(0.1, 0.9, num=3)},
    return_n_best_forecasters=2,
)

cv_random_forest = ForecastingGridSearchCV(
    forecaster=random_forest,
    cv=expanding_window_cv,
    param_grid={"estimator__max_features": np.linspace(0.1, 0.9, num=3)},
    return_n_best_forecasters=2,
)

# A combination of EnsembleForecaster(), PluginParamsForecaster() and FhPlexForecaster()
# that stacks a fixed N_BEST forecasts from both cv_linear_model and cv_random_forest and selects the 
# median for each horizon
# ...
```

**Reply by fkiraly** [reply, marked as answer]
I see - well, if you want to pool the `N_BEST` instances from both the `random_forest` and `linear_model`, you can use a union grid for `ForecastingGridSearchCV`, this works like in `sklearn`:

```python
param_grid = [
    {"estimator": linear_model, estimator__alpha": np.linspace(0.1, 0.9, num=3)},
    {"estimator": random_forest, estimator__max_features": np.linspace(0.1, 0.9, num=3)},
]

cv_automl= ForecastingGridSearchCV(
    forecaster=linear_model,
    cv=expanding_window_cv,
    param_grid=param_grid,
    return_n_best_forecasters=N_BEST,
)
```

**Reply by hliebert** [reply]
Ah, I wasn't aware you could pass a grid list like that. Thank you for the support and input! I think I can replicate everything closely enough with that.  

Specifying the grid as follows works.

```python
param_grid = [
    {"estimator": [ElasticNet()],  "estimator__alpha": np.linspace(0.1, 0.9, num=3)},
    {"estimator": [RandomForestRegressor()], "estimator__max_features": np.linspace(0.1, 0.9, num=3)},
]

cv_automl = ForecastingGridSearchCV(
    forecaster=linear_model,
    cv=expanding_window_cv,
    param_grid=param_grid,
    return_n_best_forecasters=4,
    error_score="raise",
)
```

**Reply by fkiraly** [reply]
Great to hear it works for you now!

If this solves your question in this thread, would appreciate if you could mark one of the threads/replies as the best answer - this will be pinned at the top for others who have the same question and look here.
